cpu/esp32/periph/timer.c

```diff
@@ -176,171 +159,185 @@ void IRAM hw_timer_handler(void* arg)
 
     for (unsigned dev = 0; dev < HW_TIMER_NUMOF; dev++) {
         /* iterate over all devices and check what interrupt flags are set */
-        if (timers_hw[dev].int_regs->INT_STA_REG & timers_hw[dev].int_mask) {
+
+        if (!_timers[dev].initialized) {
+            continue;
+        }
+        uint32_t int_status;
+        timer_hal_get_intr_status(&_timers[dev].hw, &int_status);
+
+        if (int_status & _timers_desc[dev].int_mask) {
             DEBUG("%s dev=%d\n", __func__, dev);
+
             /* disable alarms */
-            timers_hw[dev].regs->CONFIG_REG.LEVEL_INT_EN = 0;
-            timers_hw[dev].regs->CONFIG_REG.ALARM_EN = 0;
-            /* clear the bit in interrupt enable and status register */
-            timers_hw[dev].int_regs->INT_ENA_REG &= ~timers_hw[dev].int_mask;
-            timers_hw[dev].int_regs->INT_CLR_REG |=  timers_hw[dev].int_mask;
+            timer_hal_set_alarm_enable(&_timers[dev].hw, false);
+
+            /* disable interrupt source and clear the bit in interrupt status */
+            timer_hal_set_level_int_enable(&_timers[dev].hw, false);
+            timer_hal_intr_disable(&_timers[dev].hw);
+            timer_hal_clear_intr_status(&_timers[dev].hw);
+
             /* execute the callback function */
-            timers[dev].isr_ctx.cb(timers[dev].isr_ctx.arg, 0);
+            _timers[dev].isr_ctx.cb(_timers[dev].isr_ctx.arg, 0);
         }
     }
 
     irq_isr_exit();
 }
 
-int timer_init (tim_t dev, uint32_t freq, timer_cb_t cb, void *arg)
+int timer_init(tim_t dev, uint32_t freq, timer_cb_t cb, void *arg)
 {
+    _Static_assert(HW_TIMER_NUMOF == TIMER_NUMOF,
+                   "Number of timer descriptors does not match with TIMER_NUMOF");
+
     DEBUG("%s dev=%u freq=%" PRIu32 " cb=%p arg=%p\n",
           __func__, dev, freq, cb, arg);
 
-    CHECK_PARAM_RET (dev  <  HW_TIMER_NUMOF, -1);
-    CHECK_PARAM_RET (freq == HW_TIMER_FREQUENCY, -1);
-    CHECK_PARAM_RET (cb   != NULL, -1);
+    assert(dev  <  HW_TIMER_NUMOF);
+    assert(freq == HW_TIMER_FREQUENCY);
```

Comment:
Why is `HW_TIMER_CLK_DIV` fixed? Couldn't this be calculated at run-time to allow for arbitrary timer frequencies?

Reply:
The ESP32 implementation supports only 1 MHz timers for historical reasons, since `xtimer` worked with 1 MHz clock. It was the most simple way to use 1 MHz tick and to configure the timer clock by using the APB frequency in MHz as clock divider. It could be changed. The question is whether it should be part of this PR or a separate PR.

Reply:
Regarding the change from `CHECK_PARAM_RET` to `assert`, I am no longer sure if this change is really correct, as the API documentation says that the `timer_init` function should return -1 if the speed is not applicable or an unknown device was specified.

Reply:
Well it should be a simple change now to allow for arbitrary frequencies, so why not kill two birds with one stone here.

Reply:
Hm, not quite arbitrary. The clock divider is 16 bit. This means that the possible timer clocks depend on the APB clock and are in the range from APB clock/65,535 to APB clock. The APB clock is 80 MHz for CPU clocks of 80 MHz or higher and otherwise the CPU clock.

Theoretically, the application might select a timer clock which is outside this range. Should it throw an assert or return -1?

Reply:
For [sam0](https://github.com/RIOT-OS/RIOT/blob/master/cpu/sam0_common/periph/timer.c#L92) I went with an assert - timer config is usually static, so returning error doesn't make much sense.